api/http/api_useradm.go

```diff
@@ -576,3 +589,90 @@ func (u *UserAdmApiHandlers) GetSettingsHandler(w rest.ResponseWriter, r *rest.R
 
 	_ = w.WriteJson(settings)
 }
+
+func (u *UserAdmApiHandlers) IssueTokenHandler(w rest.ResponseWriter, r *rest.Request) {
+	ctx := r.Context()
+
+	l := log.FromContext(ctx)
+
+	var tokenRequest model.TokenRequest
+
+	if err := r.DecodeJsonPayload(&tokenRequest); err != nil {
+		rest_utils.RestErrWithLog(
+			w,
+			r,
+			l,
+			errors.New("cannot parse request body as json"),
+			http.StatusBadRequest,
+		)
+		return
+	}
+	if err := tokenRequest.Validate(u.config.TokenMaxExp); err != nil {
+		rest_utils.RestErrWithLog(
+			w,
+			r,
+			l,
+			err,
+			http.StatusBadRequest,
+		)
+		return
+	}
+
+	token, err := u.userAdm.IssuePersonalAccessToken(ctx, &tokenRequest)
+	switch err {
+	case nil:
+		writer := w.(http.ResponseWriter)
+		writer.Header().Set("Content-Type", "application/jwt")
+		_, _ = writer.Write([]byte(token))
+	case useradm.ErrTooManyTokens:
+		rest_utils.RestErrWithLog(
+			w,
+			r,
+			l,
+			err,
+			http.StatusUnprocessableEntity,
+		)
+	case useradm.ErrDuplicateTokenName:
+		rest_utils.RestErrWithLog(
+			w,
+			r,
+			l,
+			err,
+			http.StatusConflict,
+		)
+	default:
+		rest_utils.RestErrWithLogInternal(w, r, l, err)
+	}
+}
+
+func (u *UserAdmApiHandlers) GetTokensHandler(w rest.ResponseWriter, r *rest.Request) {
+	ctx := r.Context()
+	l := log.FromContext(ctx)
+	id := identity.FromContext(ctx)
+	if id == nil {
+		rest_utils.RestErrWithLogInternal(w, r, l, errors.New("identity not present"))
+		return
+	}
+
+	tokens, err := u.userAdm.GetPersonalAccessTokens(ctx, id.Subject)
+	if err != nil {
+		rest_utils.RestErrWithLogInternal(w, r, l, err)
+		return
+	}
+
+	_ = w.WriteJson(tokens)
+}
+
+func (u *UserAdmApiHandlers) DeleteTokenHandler(w rest.ResponseWriter, r *rest.Request) {
+	ctx := r.Context()
+
+	l := log.FromContext(ctx)
+
+	err := u.userAdm.DeleteToken(ctx, r.PathParam("id"))
+	if err != nil {
+		rest_utils.RestErrWithLogInternal(w, r, l, err)
+		return
+	}
+
+	w.WriteHeader(http.StatusNoContent)
+}
```

Comment:
wait this is management API. dont we need to check the identity?

Reply:
maybe I am already paranoid, but should we do `id := identity.FromContext(ctx)` and verify that given user removes his token?

Reply:
fixed